yt_dlp/extractor/nzonscreen.py

```diff
@@ -45,49 +49,109 @@ class NZOnScreenIE(InfoExtractor):
+                'duration': 1355.0,
+                'description': 'Episode 1',
+            },
+        }],
+        'info_dict': {
+            'id': 'flatmates-episode-one-1997',
+            'title': 'Flatmates - 1, First Episode',
+        },
+        'playlist_count': 5,
+    }, {
+        'url': 'https://www.nzonscreen.com/title/reluctant-hero-2008',
+        'info_dict': {
+            'id': '847f5c91af65d44b',
+            'ext': 'mp4',
+            'format_id': 'hi',
+            'height': 360,
+            'width': 640,
+            'subtitles': {
+                'en': [{'ext': 'SRT', 'data': 'md5:c2469f71020a32e55e228b532ded908f'}],
+            },
+            'title': 'Reluctant Hero (clip 1)',
+            'description': 'Part one of four from this full length documentary.',
+            'display_id': 'reluctant-hero-2008',
+            'duration': 1108.0,
+            'thumbnail': r're:https://www\.nzonscreen\.com/content/images/.+\.jpg',
+        },
+        'params': {'writesubtitles': True},
     }]
 
     def _extract_formats(self, playlist):
+        formats = []
         for quality, (id_, url) in enumerate(traverse_obj(
-                playlist, ('h264', {'lo': 'lo_res', 'hi': 'hi_res'}), expected_type=url_or_none).items()):
-            yield {
-                'url': url,
-                'format_id': id_,
-                'ext': 'mp4',
-                'quality': quality,
-                'height': int_or_none(playlist.get('height')) if id_ == 'hi' else None,
-                'width': int_or_none(playlist.get('width')) if id_ == 'hi' else None,
-                'filesize_approx': float_or_none(traverse_obj(playlist, ('h264', f'{id_}_res_mb')), invscale=1024**2),
-            }
+                playlist, ('h264', {'lo': 'lo_res', 'hi': 'hi_res', 'hd': 'hd_res'}),
+                expected_type=url_or_none).items()):
+            if traverse_obj(playlist, ('h264', f'{id_}_res_mb', {float_or_none})):
+                formats.append({
+                    'url': url,
+                    'format_id': id_,
+                    'ext': 'mp4',
+                    'quality': quality,
+                    'filesize_approx': float_or_none(traverse_obj(
+                        playlist, ('h264', f'{id_}_res_mb')), invscale=1024**2),
+                })
+        if formats:
+            formats[-1].update(traverse_obj(playlist, {
+                'height': ('height', {int_or_none}),
+                'width': ('width', {int_or_none}),
+            }))
+        return formats
+
+    def _get_subtitles(self, playinfo, video_id):
+        if caption := traverse_obj(playinfo, ('h264', 'caption_url')):
+            subtitle, urlh = self._download_webpage_handle(
+                'https://www.nzonscreen.com' + caption, video_id, 'Downloading subtitles')
+            if subtitle:
+                return {'en': [{'ext': urlhandle_detect_ext(urlh), 'data': subtitle}]}
```

Comment:
Why get `data` here instead of setting this as `url`?

Reply:
its url looks like `https://www.nzonscreen.com/captions/3367`, which probably causes YoutubeDL to fail to determine its ext when only the URL is passed.
> ```
> [info] Writing video subtitles to: test_NZOnScreen_4_ab60d7ff65270d22.en.unknown_video
> ```

the filename extension is determined by `Content-Disposition` in the response header, e.g. `Content-Disposition: attachment; filename="relucth_part1.SRT"; filename*=UTF-8''relucth_part1.SRT`. so it's necessary to make a request to determine its ext

Reply:
Would it be reasonable to hardcode extension as `srt`?

Reply:
![image](https://github.com/user-attachments/assets/2685a3b8-3b64-41bd-8e52-285b1af5ad1a)
Looking into its content, it's Content-Disposition header is misleading and i think we should hardcode vtt instead.
![image](https://github.com/user-attachments/assets/5f4d7fab-0e68-4cd4-9db3-b29b8bd74f2c)
hardcoding should be fine since the site's js is hardcoding text/vtt as well